### PR TITLE
Rule S5604 Using intrusive permissions is security-sensitive

### LINKED ISSUE #1769: Rule S5604 Using intrusive permissions is security-sensitive
Implements https://jira.sonarsource.com/browse/RSPEC-5604

For both JS/TS

Code examples: https://github.com/SonarSource/security-expected-issues/tree/master/javascript/rules/hotspots/RSPEC-5604

For the detection pattern see the comments in https://github.com/SonarSource/security-expected-issues/blob/master/javascript/rules/hotspots/RSPEC-5604/permissions.js

* Check if following syntax is used with `name` property set to the value listed in `permissions` rule parameter
```js
navigator.permissions.query({name:"geolocation"})
```

* Then for each permission from `permissions` parameter check possible APIs to use it (specified in https://github.com/SonarSource/security-expected-issues/blob/master/javascript/hotspots/RSPEC-5604/permissions.js)

* Message should depend on the reported permission/feature. e.g. `Make sure the use of the geolocation is necessary. ` if issue is reported on geolocation feature usage.
  * Camera and microphone are requested both at the same time when `getUserMedia()` method is called so in this case the message should be  `Make sure the use of the camera and microphone is necessary. `

### PR DESCRIPTION
**Pull request by andrea-guarino-sonarsource**
Fixes #1769

### CODE REVIEW

File: eslint-bridge/src/rules/no-intrusive-permissions.ts [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+            node: callee,
+          });
+        }
+      },
+    };
+  },
+};
+
+function checkPermissions(context: Rule.RuleContext, call: estree.CallExpression) {
+  const firstArg = getUniqueWriteUsageOrNode(context, call.arguments[0]);
```
Comment (guillaume-dequenne-sonarsource) [inline eslint-bridge/src/rules/no-intrusive-permissions.ts:96]: I think you can use `getValueOfExpression(context, call.arguments[0], 'ObjectExpression)` instead.

File: eslint-bridge/src/rules/no-intrusive-permissions.ts [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+    isIdentifier(property, ...secondProperty)
+  );
+}
+
+function hasNamePropertyWithPermission(
+  prop: estree.Property | estree.SpreadElement,
+  context: Rule.RuleContext,
+) {
+  if (prop.type === 'Property' && isIdentifier(prop.key, 'name')) {
+    const value = getUniqueWriteUsageOrNode(context, prop.value);
```
Comment (guillaume-dequenne-sonarsource) [inline eslint-bridge/src/rules/no-intrusive-permissions.ts:124]: Same here, you can use `getValueOfExpression(context, prop.value, 'Literal')`

**Review by guillaume-dequenne-sonarsource — CHANGES_REQUESTED** [review]
I left a few minor comments and there's probably an issue message that should be updated, but looks good otherwise.

File: eslint-bridge/src/rules/no-intrusive-permissions.ts [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+  },
+};
+
+function checkPermissions(context: Rule.RuleContext, call: estree.CallExpression) {
+  const firstArg = getUniqueWriteUsageOrNode(context, call.arguments[0]);
+  if (firstArg.type === 'ObjectExpression') {
+    const nameProp = firstArg.properties.find(prop => hasNamePropertyWithPermission(prop, context));
+    if (nameProp) {
+      context.report({
+        message: 'Make sure the use of the geolocation is necessary.',
```
[the remark this answers is not in the document]
Comment (andrea-guarino-sonarsource) [reply eslint-bridge/src/rules/no-intrusive-permissions.ts:101]: Indeed, I forgot to parametrize the message. thank you!

File: eslint-bridge/src/rules/no-intrusive-permissions.ts [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+        ) {
+          context.report({
+            message: 'Make sure the use of the geolocation is necessary.',
+            node: callee,
+          });
+          return;
+        }
+        if (
+          context.options.includes('camera') &&
+          context.options.includes('microphone') &&
```
[the remark this answers is not in the document]
Comment (andrea-guarino-sonarsource) [reply eslint-bridge/src/rules/no-intrusive-permissions.ts:53]: In the ticket description, there is the following:
> Camera and microphone are requested both at the same time when getUserMedia() method is called so in this case the message should be Make sure the use of the camera and microphone is necessary.

I'm not sure we should report only when they are both included or one is enough.
@eric-therond-sonarsource wdyt?

[the remark this answers is not in the document]
Comment (andrea-guarino-sonarsource) [reply eslint-bridge/src/rules/no-intrusive-permissions.ts:53]: > Also I see on the examples that there might be a constraints argument on the navigator.mediaDevices.getUserMedia call, with audio and video properties. Can this be relevant to when we should decide to report? (I assume it's not worth tracking though - just took the opportunity to address it explicitly).

As far as I've understood it's not relevant to raise the issue. @eric-therond-sonarsource can you please clarify?

[the remark this answers is not in the document]
Comment (eric-therond-sonarsource) [reply eslint-bridge/src/rules/no-intrusive-permissions.ts:53]: About `constraints` argument you are right, good catch, 
the constraints object is mandatory, if not passed to the `getUserMedia` method, both audio and video will not be requested and when `audio` property or `video` property different from `false` then the corresponding media will be requested:
```
navigator.mediaDevices.getUserMedia({ audio: true, video: true }) // Sensitive for camera and microphone (if both permissions have been defined by the user in the parameter of this custom rule)
```

[the remark this answers is not in the document]
Comment (eric-therond-sonarsource) [reply eslint-bridge/src/rules/no-intrusive-permissions.ts:53]: And so the message suggested on  he ticket is wrong:
> Camera and microphone are requested both at the same time when getUserMedia() method is called so in this case the message should be Make sure the use of the camera and microphone is necessary.

we should have this message "_Make sure the use of the [camera (or|and) microphone] is/are necessary_." depending on values of the audio and video properties on the constraints object.

[the remark this answers is not in the document]
Comment (eric-therond-sonarsource) [reply eslint-bridge/src/rules/no-intrusive-permissions.ts:53]: I did an additional test if `constraints` object is passed to `getUserMedia` but `audio` property or `video` property is missing then the corresponding media will not be requested by the browser:

```
navigator.mediaDevices.getUserMedia({ video: true }) // Sensitive for camera only (if camera permission has been defined by the user in the parameter of this custom rule)
```

[the remark this answers is not in the document]
Comment (andrea-guarino-sonarsource) [reply eslint-bridge/src/rules/no-intrusive-permissions.ts:53]: ok thanks! I will apply those changes!

**Review by guillaume-dequenne-sonarsource — APPROVED** [review]
LGTM

**Pull request merged**